### PR TITLE
Issue 4356: publish queue notification

### PR DESCRIPTION
**Pull request by Niicck**
Issue: https://github.com/cityofaustin/techstack/issues/4356
Related to: https://github.com/cityofaustin/publisher/pull/13/files

This ended up being a big one. The amount of code changed across 2 repositories is overwhelming to look at without context. So here's a play-by-play walkthrough of what's been updated and why. Hopefully this will make code review easier to do.

- [ ] Added 4 new publish_* fields to `joplin/pages/base_page/models.py` to handle the real "Publishing" state. Things are no longer just "Live" or "Draft". https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L51
- [ ] The logic to set the new page_status is in the `status_string` method in `joplin/pages/base_page/models.py` https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L162
- [ ] The template to render those new status_string options is in `joplin/templates/wagtailadmin/shared/page_status_tag.html` https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/templates/wagtailadmin/shared/page_status_tag.html
- [ ] We had to basically completely redo how we're collecting pages to send to the Publisher. We're now sending a lot more data than just the page_id. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/base/signals/janis_build_triggers.py
- [ ] That data gets sent to the publisher. https://github.com/cityofaustin/publisher/pull/13/files
- [ ] The page data is received by the publisher's newly updated publish_request_handler. https://github.com/cityofaustin/publisher/blob/4356-queue/src/handlers/publish_request_handler.py
- [ ] The publish_request_handler now has extensive tests to handle edge cases and errors. https://github.com/cityofaustin/publisher/blob/4356-queue/src/tests/test_publish_request_handler.py
- [ ] While our code is building, we now display fancy badges for our "Publishing" and "Unpublishing" page statuses. And if you click on them, a nice informational modal pops up! And that was really the goal for this issue. Everything else in these 2 PRs are just means to an end of displaying this 15 line html file. What is life. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/templates/joplin_UI/pageStatusModals/publishing.html
- [ ] At the end of a successful build in Publisher, it will send a response back to Joplin. https://github.com/cityofaustin/publisher/blob/4356-queue/src/handlers/commands/send_publish_succeeded_message.py
- [ ] Joplin receives this message. We've installed the django-rest-framework and an api-key library to authenticate this endpoint. That way, only the Publisher (with a valid API key) will be able to access this open endpoint. https://github.com/cityofaustin/joplin/blob/4356-queue/joplin/base/views/publish_succeeded.py

Bonus things:
- [ ] Our PR Heroku postgres databases will now use version 12, just like staging and prod.
- [ ] I added uglify compression to our production webpack builds for Joplin. It appears to have had no impact. To be continued.
- [ ] new_page_from_modal now uses a `@require_POST` decorator instead of an "if request.POST" statement inside of it. That's why it looks like there are a bunch of changes. That code basically hasn't changed, it's just unindented.
- [ ] all fixtures now have revisions, and "live" fixtures are also "published"

TODO:
- [ ] Need to ask design where this little banner goes: https://app.abstract.com/projects/bbe34f6b-7080-4dcd-9fa9-d8e30d8f6128/branches/master/commits/f90d313fba24e354093e1baffe92abaceac1fd79/files/a92ebeb2-03b6-44cb-8376-a1b7230da8bd/layers/9073443f-4254-4cf8-a52b-11e8ccbf49b8?collectionId=acbfaded-c24c-4ea0-b8de-b87e4758f854&collectionLayerId=8d4c885d-16de-4750-85f9-8a05b946d520&mode=build

## How to Test
Go to https://joplin-pr-4356-queue.herokuapp.com/admin/pages/search/. Feel free to go crazy. Publish stuff, unpublish stuff. You should see the correct page_status badges appear in accordance with the rules set forth here: https://github.com/cityofaustin/joplin/blob/9be665c63d0f7f79649f666bdfb08673bf7258d4/joplin/pages/base_page/models.py#L162

We built a distinct Publisher environment that has the new changes from the Publisher PR for this issue. You can track your builds by going to this dynamodb table: https://console.aws.amazon.com/dynamodb/home?region=us-east-1#tables:selected=coa_publisher_4356-queue;tab=items

### CODE REVIEW

File: joplin/settings.py [hunk 1]
```diff
@@ -500,6 +510,10 @@ def show_toolbar(request):
         },
     },
 }
+# Send email notifications of errors for PR build
+# TODO: this can be managed in a cleaner way
+if os.getenv("CIRCLE_BRANCH") == "4356-queue":
+    ADMINS = [('Nick', 'nick.ivons@austintexas.gov')]
```
Comment (Niicck) [inline joplin/settings.py:516]: I'll clean this up before merging

File: README.md [hunk 2]
```diff
@@ -358,6 +358,10 @@ The migration process currently consists of 3 commands:
 
 -   ssh into a container that's deployed on Heroku
     -   `heroku run bash -a $APPNAME`
+-   Copy data from one app into another
+    -   `heroku pg:copy source-app::DATABASE_URL DATABASE_URL -a destination-app`
+    - Plug in the actual names of your "source-app" and "destination-app".
+    - Literally use "DATABASE_URL" in your command.
```
Comment (chiaberry) [inline README.md:364]: thank you!

File: joplin/base/views/new_page_from_modal.py [hunk 3]
```diff
@@ -25,98 +26,100 @@ def import_page_from_url(url, jwt_token):
     return page.id
 
 
+@require_POST
 def new_page_from_modal(request):
     user_perms = UserPagePermissionsProxy(request.user)
     if not user_perms.can_edit_pages():
         raise PermissionDenied
 
-    if request.method == 'POST':
```
Comment (chiaberry) [inline joplin/base/views/new_page_from_modal.py:33]: out of curiosity, what other request.method would be /could have been passed into here?

File: joplin/base/views/new_page_from_modal.py [hunk 4]
```diff
@@ -25,98 +26,100 @@ def import_page_from_url(url, jwt_token):
...
 
-        # Add it as a child of home
-        home = HomePage.objects.first()
-        home.add_child(instance=page)
+    # Save our draft
+    page.save_revision()
 
-        # Save our draft
-        page.save_revision()
-        page.unpublish()  # Not sure why it seems to go live by default
```
Comment (chiaberry) [inline joplin/base/views/new_page_from_modal.py:83]: why dont we need to unpublish anymore?

Comment (Niicck) [reply joplin/base/views/new_page_from_modal.py:83]: Because now we're setting new pages in create_content_modal to be not live by default.
```
# Pages will go live by default if we don't explicitly specify live=False
    data = {
        'live': False
    }
    data['title'] = body['title']
    data['owner'] = request.user
```

File: joplin/base/views/new_page_from_modal.py (hunk 3 above)
Comment (Niicck) [reply joplin/base/views/new_page_from_modal.py:33]: Nothing logical. We aren't making any GET, DELETE, or PUT requests to new_page_from_modal.

File: joplin/pages/base_page/models.py [hunk 5]
```diff
@@ -48,6 +48,17 @@ class JanisBasePage(Page):
...
 
+    # Fields used for handling Publishing status
+    # pk for our publish request in Publisher dynamodb
+    publish_request_pk = models.TextField(blank=True, null=True)
+    # sk for our publish request in Publisher dynamodb
+    publish_request_sk = models.TextField(blank=True, null=True)
+    # Indicated whether a publish_request for this page been submitted to the Publisher, and we are waiting for it to finish being processed.
+    publish_request_enqueued = models.BooleanField(default=False)
+    # Has this page been published by Publisher? A "live" page may not necessarily be published to our frontend yet.
+    published = models.BooleanField(default=False, blank=True, null=True)
```
Comment (chiaberry) [inline joplin/pages/base_page/models.py:59]: a live page may not necessarily be published because of that liminal state of "currently being published" or heaven forbid, a page that is marked as live but b/c of error didnt actually get published?

Comment (chiaberry) [reply joplin/pages/base_page/models.py:59]: yup, i just read further down and answered my own question

Comment (Niicck) [reply joplin/pages/base_page/models.py:59]: If there is an error with publishing and it didn't get published, then it will be in a "Publishing" status. This is by design until there's some more specific error messaging we want to add.

This is actually a good approach. Because the Publisher is smart. When we deploy the needed fix to Joplin/Janis/Publisher, the Publisher knows to sweep up all previously failed publish_requests and retry them. And our views/publish_succeeded endpoint will iterate through every page in the build and update pages that "triggered_builds" to set their status as no longer publishing.

File: Pipfile [hunk 6]
```diff
@@ -49,6 +49,8 @@ snapshottest = "*"
 gql = "*"
 pyhumps = "*"
 django-graphql-jwt = "*"
+djangorestframework = "*"
```
Comment (easherma) [inline Pipfile:52]: Just curious, since DRF is already a dependency of wagtail I would've expected you to be able to work with it without adding it as a direct dependency, was that not the case?

File: .circleci/scripts/build_heroku_infrastructure.sh [hunk 7]
```diff
@@ -8,7 +8,7 @@ log() { log_base "build_heroku_app" "$1" "$2"; }
 
 # Attaches postgresql database to heroku application
 function attach_heroku_database {
-    heroku addons:create heroku-postgresql:hobby-dev --version=10 --app $APPNAME
+    heroku addons:create heroku-postgresql:hobby-dev --version=12 --app $APPNAME
```
Comment (briaguya) [inline .circleci/scripts/build_heroku_infrastructure.sh:11]: hooray!

File: README/testing.md [hunk 8]
```diff
@@ -9,6 +9,23 @@ We now have automated tests! They get run on every circleci build during the "te
...
+
+Run one test case:
+`pipenv run pytest pages/information_page/tests.py::test_import_from_page_dictionary_twice_different_revisions`
+
+Reuse database:
+`--reuse-db`
+
+Log print statements:
+`-s`
+
```
Comment (briaguya) [inline README/testing.md:28]: hooray for documentation!

File: joplin/base/signals/janis_build_triggers.py [hunk 9]
```diff
@@ -1,131 +1,212 @@
...
+        "published" - publish action for itself and other pages
+        "unpublished" - this page triggered the unpublish action for itself and other pages
+        "updated_by_snippet" - this page was updated by a snippet being saved or deleted
+        "saved" - a snippet is being saved, has no impact on Janis itself, but it could result in pages getting "secondary_publish_by_snippet"
+        "deleted" - a snippet is being deleted, has no impact on Janis itself, but it could result in pages getting "secondary_publish_by_snippet"
+    is_page: Boolean, is it a page as opposed to a snippet?
+    content_type: String, which specific content_type (this is more for logging)
+    author: Int, the id of the author of the latest revision
+}
+TODO: Figure out what format global_id should be in order to run queries with it.
```
Comment (briaguya) [inline joplin/base/signals/janis_build_triggers.py:40]: i think these are less complex in grapple

File: joplin/base/signals/janis_build_triggers.py [hunk 10]
```diff
@@ -1,131 +1,212 @@
...
+    }
+
+
+def collect_pages(primary_page, action):
     # does this work on page deletion? pages arent deleted right, just unpublished?
     """
-    :param instance: the page that has been altered
-    :return: an array of global page ids -- note this may need to be a dictionary of page ids and type?
+    :param primary_page: the page that triggered the publish/unpublish
+    :return: pages: an array of the data from all pages impacted by the publish/unpublish event
```
Comment (briaguya) [inline joplin/base/signals/janis_build_triggers.py:65]: this is awesome

File: joplin/base/tests/test_incremental_builds.py [hunk 11]
```diff
@@ -18,7 +18,7 @@
 
 from wagtail.documents.models import Document
 
-from base.signals.janis_build_triggers import collect_pages, collect_pages_snippet
+from base.signals.janis_build_triggers import collect_pages, collect_pages_from_snippet
```
Comment (briaguya) [inline joplin/base/tests/test_incremental_builds.py:21]: much better name!

File: joplin/base/views/new_page_from_modal.py [hunk 12]
```diff
@@ -25,98 +26,100 @@ def import_page_from_url(url, jwt_token):
     return page.id
 
 
+@require_POST
```
Comment (briaguya) [inline joplin/base/views/new_page_from_modal.py:29]: nice!

File: joplin/base/signals/janis_build_triggers.py (hunk 9 above)
Comment (Niicck) [reply joplin/base/signals/janis_build_triggers.py:40]: Ah good to know

File: joplin/base/signals/janis_build_triggers.py (hunk 10 above)
Comment (Niicck) [reply joplin/base/signals/janis_build_triggers.py:65]: It's basically not used for anything now, but it sets us up in a better place for incremental builds later.

File: Pipfile (hunk 6 above)
Comment (Niicck) [reply Pipfile:52]: I just tried running it without using it as a direct dependency. And it does not work.
`ModuleNotFoundError: No module named 'rest_framework'`
